kubeadm: Support custom env in control plane component

chendave opened the pull request:
/kind feature

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
```
There are some cases,  the custom environments such as `GOMAXPROCS` could be consumed by the control plane component.
This change supports the configuration of custom environments in control plane components with
the `ClusterConfiguration.Etcd.Local.ExtraEnvs`, `cfg.APIServer.ExtraEnvs`, `cfg.ControllerManager.ExtraEnvs`, `cfg.Scheduler.ExtraEnvs`.
```

```
Fixed some typos in the v1beta4/doc.go
```
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
xref: https://github.com/kubernetes/kubeadm/issues/2890

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

chendave committed:
kubeadm: Support custom env in control plane component

Signed-off-by: Dave Chen <dave.chen@arm.com>

chendave commented:
cc other members signed with the v1beta4. 

/cc @neolit123 @ruquanzhao 

I tested this against v1beta3, and it works, set the `GOMAXPROCS` in the config file and apiserver show me that env is set inside the container.

> I0626 09:17:44.842730       1 server.go:167] "Golang settings" GOGC="" GOMAXPROCS="10" GOTRACEBACK=""

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -150,6 +150,10 @@ type ControlPlaneComponent struct {
 	// ExtraVolumes is an extra set of host volumes, mounted to the control plane component.
 	ExtraVolumes []HostPathMount
+
+	// ExtraEnvs is an extra set of envs to pass to the control plane component, e.g. GOMAXPROCS.
```
```suggestion
	// ExtraEnvs is an extra set of environment variables to pass to the control plane component
```

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta3/conversion.go:
```diff
@@ -34,3 +35,27 @@ func Convert_v1beta3_InitConfiguration_To_kubeadm_InitConfiguration(in *InitConf
 }
+
+// Convert_v1beta3_ControlPlaneComponent_To_kubeadm_ControlPlaneComponent is required due to the missing ControlPlaneComponent.ExtraEnvs in v1beta3.
+// TODO: remove this with v1beta3 https://github.com/kubernetes/kubeadm/issues/2890
```
the todo is not needed, given the package folder will be removed as a whole. same in other new conversion functions.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -262,6 +266,10 @@ type LocalEtcd struct {
 
+	// ExtraEnvs is an extra set of envs to pass to the control plane component, e.g. GOMAXPROCS.
+	// +optional
+	ExtraEnvs []v1.EnvVar `json:"extraEnvs,omitempty"`
```
```suggestion
	ExtraEnvs []v1.EnvVar
```

internal structs must be without json tags

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go:
```diff
@@ -150,6 +150,10 @@ type ControlPlaneComponent struct {
 	// +optional
 	ExtraVolumes []HostPathMount `json:"extraVolumes,omitempty"`
+
+	// ExtraEnvs is an extra set of envs to pass to the control plane component, e.g. GOMAXPROCS.
```
use adjusted comment from internal type

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go:
```diff
@@ -282,6 +286,10 @@ type LocalEtcd struct {
 	// +optional
 	ExtraArgs map[string]string `json:"extraArgs,omitempty"`
 
+	// ExtraEnvs is an extra set of envs to pass to the control plane component, e.g. GOMAXPROCS.
```
ditto

neolit123 commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -64,7 +64,7 @@ func GetStaticPodSpecs(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmap
 			StartupProbe:    staticpodutil.StartupProbe(staticpodutil.GetAPIServerProbeAddress(endpoint), "/livez", int(endpoint.BindPort), v1.URISchemeHTTPS, cfg.APIServer.TimeoutForControlPlane),
 			Resources:       staticpodutil.ComponentResources("250m"),
-			Env:             kubeadmutil.GetProxyEnvVars(),
+			Env:             append(cfg.APIServer.ExtraEnvs, kubeadmutil.GetProxyEnvVars()...),
```
this is not merging the proxy and user vars?

they will be: FOO, BAR, FOO

i think we should make the user vars take priority / override and remove duplicates

neolit123 commented on cmd/kubeadm/app/phases/controlplane/manifests_test.go:
```diff
@@ -51,6 +53,9 @@ func TestGetStaticPodSpecs(t *testing.T) {
 	cfg := &kubeadmapi.ClusterConfiguration{
 		KubernetesVersion: "v1.9.0",
+		Scheduler: kubeadmapi.ControlPlaneComponent{ExtraEnvs: []v1.EnvVar{
+			{Name: "Foo", Value: "Bar"},
```
we should have override / de-dup tests

neolit123 commented on cmd/kubeadm/app/phases/controlplane/manifests_test.go:
```diff
@@ -28,8 +28,10 @@ import (
+	"github.com/google/go-cmp/cmp"
 	"github.com/lithammer/dedent"
 
+	v1 "k8s.io/api/core/v1"
```
move to line below, where the api-machinery import group is

neolit123 commented on cmd/kubeadm/app/phases/controlplane/manifests_test.go:
```diff
@@ -28,8 +28,10 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/google/go-cmp/cmp"
```
we already import reflect. maybe we can use it instead of the extra import of cmp

neolit123 replied:
reflect should be fine for slices, but i don't insist on not using cmp

chendave replied:
Gotcha, I will switch to reflect in the next iteration.

chendave replied:
@neolit123 have switched to reflect.

neolit123 commented on cmd/kubeadm/app/phases/etcd/local_test.go:
```diff
@@ -27,8 +27,10 @@ import (
 	"sort"
 	"testing"
 
```
same import comment about reflect

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -150,6 +150,10 @@ type ControlPlaneComponent struct {
 	// ExtraVolumes is an extra set of host volumes, mounted to the control plane component.
 	ExtraVolumes []HostPathMount
+
+	// ExtraEnvs is an extra set of envs to pass to the control plane component, e.g. GOMAXPROCS.
```
are we ok with the name extraenvs? @SataQiu @pacoxu

SataQiu replied:
seems ok

neolit123 reviewed (commented):
thanks @chendave 
added some comments

SataQiu commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -76,7 +76,7 @@ func GetStaticPodSpecs(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmap
 			StartupProbe:    staticpodutil.StartupProbe(staticpodutil.GetControllerManagerProbeAddress(cfg), "/healthz", kubeadmconstants.KubeControllerManagerPort, v1.URISchemeHTTPS, cfg.APIServer.TimeoutForControlPlane),
 			Resources:       staticpodutil.ComponentResources("200m"),
-			Env:             kubeadmutil.GetProxyEnvVars(),
+			Env:             append(cfg.ControllerManager.ExtraEnvs, kubeadmutil.GetProxyEnvVars()...),
```
ditto

SataQiu commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -87,7 +87,7 @@ func GetStaticPodSpecs(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmap
 			StartupProbe:    staticpodutil.StartupProbe(staticpodutil.GetSchedulerProbeAddress(cfg), "/healthz", kubeadmconstants.KubeSchedulerPort, v1.URISchemeHTTPS, cfg.APIServer.TimeoutForControlPlane),
 			Resources:       staticpodutil.ComponentResources("100m"),
-			Env:             kubeadmutil.GetProxyEnvVars(),
+			Env:             append(cfg.Scheduler.ExtraEnvs, kubeadmutil.GetProxyEnvVars()...),
```
ditto

chendave committed:
kubeadm: fix fuzzing for ExtraEnvs defined in the control plane component

Signed-off-by: Dave Chen <dave.chen@arm.com>

neolit123 commented on cmd/kubeadm/app/util/env.go:
```diff
@@ -41,3 +41,17 @@ func GetProxyEnvVars() []v1.EnvVar {
 	return envs
 }
+
+// MergedEnv merge the extraEnvs with proxyEnv, extraEnvs take precedence over the proxyEnv.
```
func should be called with a verb. MergeEnv

neolit123 replied:
it should be a generic, variadic argument function that takes a slice if env slices and makes the last slices override / take precedence.

chendave replied:
Done, pls let me know if the current version address your comment.

chendave commented on cmd/kubeadm/app/phases/controlplane/manifests_test.go:
```diff
@@ -71,19 +76,30 @@ func TestGetStaticPodSpecs(t *testing.T) {
-
+				if tc.env != nil {
+					envs := []v1.EnvVar{}
+					// Just compare the env that testcase set as the proxy env might be set an arbitrary value on the test node.
```
we must remove the "_proxy" env here, as the test node has the "no_proxy"env configured.

neolit123 replied:
this sounds like it needs refactor to not have tests that depend on the host env. i am talking without checking, but the host env could be passed only on actual runs / outside of tests.

chendave replied:
I am not sure if am following here,  correct me pls, how the host env is set seems not matter here as we just need to make sure the env set in the clusterCfg is parsed correctly.

chendave replied:
@neolit123 pls see the test result without the tweak here: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/118867/pull-kubernetes-unit/1673936678231740416

neolit123 replied:
seems to me we should refactor our test here not to get:
>  [{no_proxy 127.0.0.1,localhost nil}

if i am not mistaken these come from the host?

chendave replied:
yes, this comes from the host. I will leave a TODO here so that we can revisit it later if we have a better way to tackle it.

neolit123 replied:
i think we can fix it here as it's relatively simple.

- make  GetStaticPodSpecs accept a []v1.EnvVar called proxyEnvs
- if the slice is nil populate it with GetProxyEnvVars() and use it in manifest construction
- if the slice is non-nil use it's value instead of GetProxyEnvVars()
- then in this unit test just pass an empty slice

does it make sense to you?

chendave replied:
Done, thanks for the guidance!

neolit123 commented on cmd/kubeadm/app/util/env_test.go:
```diff
@@ -0,0 +1,76 @@
+)
+
+func TestMergeEnv(t *testing.T) {
+	proxyEnv := []v1.EnvVar{}
```
we can call it e.g. baseEnv. it doesn't have to be proxy, always

chendave replied:
yep!

neolit123 commented on cmd/kubeadm/app/util/env.go:
```diff
@@ -41,3 +41,18 @@ func GetProxyEnvVars() []v1.EnvVar {
 	return envs
 }
+
+// MergeEnv merges the value of the environment variables, the value defined in the last slice overrides the previous one.
```
```suggestion
// MergeEnv merges values of environment variable slices. The values defined in later slices overwrite values in previous ones.
```

chendave replied:
Done

pacoxu reviewed (commented):
/lgtm
a release note is needed.

We may add an example to the website for this change.

pacoxu commented:
Live v1beta3, comment here needs an update. Or we may update it later.

https://github.com/kubernetes/kubernetes/blob/8d4479a58f0b1864bdc0ec386e045239b89d15d2/cmd/kubeadm/app/apis/kubeadm/v1beta3/doc.go#L25-L27

There is already a TODO.
https://github.com/kubernetes/kubernetes/blob/8d4479a58f0b1864bdc0ec386e045239b89d15d2/cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go#L25-L27

neolit123 commented:
> a release note is needed.

no, as discussed we will add them all at once near beta4 release

pacoxu commented:
> > a release note is needed.
> 
> 
> 
> no, as discussed we will add them all at once near beta4 release

got it

chendave commented:
the change since the last push:
https://github.com/kubernetes/kubernetes/compare/079c1c66b624c9beab99085abf27a59381bdacaa..87940cc7294542a628733cac8a2a60c841da54cb

> We may add an example to the website for this change.

will do in the 1.29 release cycle.

> please add note in v1beta4 doc.go where the todo is.

done

chendave commented:
@neolit123 pls let me know is there anything more need to be addressed here.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -20,10 +20,13 @@ limitations under the License.
 // A list of changes since v1beta3:
+//
 //   - TODO https://github.com/kubernetes/kubeadm/issues/2890
+//   - Support custom environment in control plane components
```
```suggestion
//   - Support custom environment variables in control plane components under `ClusterConfiguration`.
```

chendave replied:
Done

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -20,10 +20,13 @@ limitations under the License.
+//
 //   - TODO https://github.com/kubernetes/kubeadm/issues/2890
+//   - Support custom environment in control plane components
+//     Custom environments could be configured with `ClusterConfiguration.Etcd.Local.ExtraEnvs`, `cfg.APIServer.ExtraEnvs`, `cfg.ControllerManager.ExtraEnvs`, `cfg.Scheduler.ExtraEnvs`.
```
```suggestion
//     Use `APIServer.ExtraEnvs`, `ControllerManager.ExtraEnvs`, `Scheduler.ExtraEnvs`,  `Etcd.Local.ExtraEnvs`.
```

chendave replied:
Done

neolit123 reviewed (commented):
added one nit about the doc.go.

/approve
leaving LGTM to another pair of eyes.
note: ideally we should have multiple reviewers on all important PRs (e.g. API).

chendave commented:
/assign @pacoxu @SataQiu 

for a double check.

pacoxu commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -48,9 +48,12 @@ func CreateInitStaticPodManifestFiles(manifestDir, patchesDir string, cfg *kubea
 	mounts := getHostPathVolumesForTheControlPlane(cfg)
+	if proxyEnvs == nil {
+		proxyEnvs = kubeadmutil.GetProxyEnvVars()
+	}
```
The code of `GetStaticPodSpecs` is a little tricky here.

It checked if the last arg is nil. If nil, the function will add proxy env. If not nil(including empty), it will not add proxy env. The empty case is used in the test only. https://github.com/kubernetes/kubernetes/pull/118867/files#diff-1937a20cf97bf14e2aea8ae19d61d935536d2c903b49adf4f4502fcfd89a5296R61.

No place will really pass a real env into the function.

```
func GetStaticPodSpecs(cfg *kubeadmapi.ClusterConfiguration, endpoint *kubeadmapi.APIEndpoint) map[string]v1.Pod {
	// Get the required hostpath mounts
	mounts := getHostPathVolumesForTheControlPlane(cfg)
	proxyEnvs := kubeadmutil.GetProxyEnvVars()

```

The test below may just check the last env of the container[0].

Just for nit.

chendave replied:
this was discussed here: https://github.com/kubernetes/kubernetes/pull/118867#discussion_r1249195215, I am fine with either of them but a little in favor of the current approach (we cannot guarantee extraEnv is always the last env, and actually it's a slice and we cannot inspect easily how many envs the user sets), the original approach is filter the proxy env out, @neolit123 suggest me to remove the dep over the host.

pacoxu replied:
Sorry to miss the comments above 😄. 

It seems to be on test purposes to not be affected by host env. 

It is fine.

chendave commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/doc.go:
```diff
@@ -20,10 +20,13 @@ limitations under the License.
+//
 //   - TODO https://github.com/kubernetes/kubeadm/issues/2890
+//   - Support custom environment variables in control plane components under `ClusterConfiguration`.
+//     Use `APIServer.ExtraEnvs`, `ControllerManager.ExtraEnvs`, `Scheduler.ExtraEnvs`,  `Etcd.Local.ExtraEnvs`.
```
likes like to me addons are not the resource  maintained by kubeadm. @neolit123 @SataQiu  anything to comment here?

neolit123 replied:
some kubeadm addon context.

in the beginning we wanted to make coredns and kube-proxy more tightly coupled with the kubeadm API and customization, e.g. ExtraVolumes, patches, etc. but instead we started discussions that we want to make these addons external and maintain their manifests and configuration as external, where e.g. an addon manager just pulls the configuration for the adon. same addon manager tells where kubeadm can download container images, etc.

this work never happened and we never arrived on a good design solution. as a result, we don't want to integrate these addons more tightly with kubeadm, and just tell users if they want more customization they can "skip" these addons and deploy their own.

chendave replied:
thanks for the context!

chendave commented on cmd/kubeadm/app/util/env_test.go:
```diff
@@ -0,0 +1,76 @@
+	}{
+		{
+			name: "normal case without duplicated env",
+			proxyEnv: []v1.EnvVar{
```
+1, this is just a testcase.

chendave reviewed (commented):
@pacoxu I replied all your comments, pls let me know is that make sense to you, thanks!

pacoxu commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go:
```diff
@@ -282,6 +286,10 @@ type LocalEtcd struct {
 
+	// ExtraEnvs is an extra set of environment variables to pass to the control plane component.
+	// +optional
+	ExtraEnvs []corev1.EnvVar `json:"extraEnvs,omitempty"`
```
As below, the `ExtraEnvs` will overwrite the host `ProxyEnv` on purpose.

Nit: to mention it in the comments?

neolit123 replied:
well, what we have here is corev1.EnvVar, which means users can actually use Value but also ValueFrom:
https://pkg.go.dev/k8s.io/api/core/v1#EnvVar

and they can do things like this:
https://kubernetes.io/docs/tasks/inject-data-application/environment-variable-expose-pod-information/#use-pod-fields-as-values-for-environment-variables

so technically our ExtraEnvs will support the downward API.

neolit123 replied:
> As below, the `ExtraEnvs` will overwrite the host `ProxyEnv` on purpose if the same(only for `*_proxy`).
> 
> Nit: to mention it in the comments?

hmm, that's a good idea, in case users ask.

@chendave we can use this, WDYT?

>> // ExtraEnvs is an extra set of environment variables to pass to the control plane component.
>> // Environment variables passed using ExtraEnvs will override any existing environment variables,
>> // or *_proxy environment variables that kubeadm adds by default.

pacoxu commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -262,6 +267,10 @@ type LocalEtcd struct {
 
+	// ExtraEnvs is an extra set of environment variables to pass to the control plane component.
+	// +optional
+	ExtraEnvs []v1.EnvVar
```
Comments are not updated. https://github.com/kubernetes/kubernetes/pull/118867/files#diff-a919b4d8ff1a3b03bb5691a0725e0683ff5e3fa4e12c4f81690ed46b156816d7R290 too,

Overall LGTM.

chendave replied:
@pacoxu this is etcd, there is no such proxy env passed to etcd compared with `ControlPlaneComponent`, and we haven't override anything like what we does for `ControlPlaneComponent`.

chendave replied:
https://github.com/kubernetes/kubernetes/pull/118867/files#diff-509876ddb9971f8787c885617ebaabe5da81755c5f1cddc1e2877af0d47928b2R227

neolit123 replied:
/hold

please add it everywhere. doesn't matter if we skip proxy.

actually why are we not adding proxy envs to etcd? isn't that a bug?

pacoxu replied:
Generally etcd will not access outside, only peers should be accessed by etcd. And only local apiserver can access etcd.

For apiserver, it may access webhooks. Scheduler also can have webhook.

neolit123 replied:
hm, maybe a peer would not be able to reach a peer because of proxy.

chendave replied:
there is no such a request for passing the proxy env so far, we can add it when there is a clear usecase,  and this is not a API change, it will be easy to address then.

chendave replied:
added the comment for all the occurrences! @pacoxu @neolit123 looking for LGTM again.

pacoxu replied:
A corner case that is nice to support😄. Either is OK with me for this point.

chendave commented:
@pacoxu anything more to add?

pacoxu commented:
/lgtm

neolit123 commented:
/lgtm
/hold cancel

pacoxu commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go:
```diff
@@ -150,6 +150,11 @@ type ControlPlaneComponent struct {
+	// ExtraEnvs is an extra set of environment variables to pass to the control plane component.
+	// Environment variables passed using ExtraEnvs will override any existing environment variables, or *_proxy environment variables that kubeadm adds by default.
+	// +optional
+	ExtraEnvs []corev1.EnvVar `json:"extraEnvs,omitempty"`
```
This caused the import of apis/core/v1 in generated.defaults below.

Should we avoid that or how to avoid the defaulting logic?

neolit123 replied:
we already had this import:
https://github.com/chendave/kubernetes/blob/c68a6b07456301f6a8102c01e99ba8a53bc9bf4f/cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go#L20

https://github.com/chendave/kubernetes/blob/c68a6b07456301f6a8102c01e99ba8a53bc9bf4f/cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go#L226
https://github.com/chendave/kubernetes/blob/c68a6b07456301f6a8102c01e99ba8a53bc9bf4f/cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go#L244
https://github.com/chendave/kubernetes/blob/c68a6b07456301f6a8102c01e99ba8a53bc9bf4f/cmd/kubeadm/app/apis/kubeadm/v1beta4/types.go#L441

pacoxu replied:
So this not results to the `import` in generated_default.go?

https://github.com/kubernetes/kubernetes/pull/118867/files#r1321045615

pacoxu commented on cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go:
```diff
@@ -23,6 +23,7 @@ package v1beta4
 
 import (
 	runtime "k8s.io/apimachinery/pkg/runtime"
+	v1 "k8s.io/kubernetes/pkg/apis/core/v1"
```
As I tested, this is the key import that is related.

neolit123 replied:
@pacoxu as mentioned the kubernetes/pkg import is _not allowed_ in kubeadm!

@chendave we must find out how to solve it. maybe it's a generator bug.
if we cannot solve it we can try avoiding the defaulting `[]...{}`

if we are out of ideas, we must revert the PR.

SataQiu replied:
We must drop this `import` first.
In the past, we have made a lot of efforts to do this, and we cannot go back.
But I'm confused, why didn't the [.import-restrictions](https://raw.githubusercontent.com/kubernetes/kubernetes/master/cmd/kubeadm/.import-restrictions) prevent PRs from being merged?

neolit123 replied:
perhaps the import-boss tool is not running or buggy.
it can be tested with a simple PR for cmd/kubeadm or other packages that have an .import-restrictions file.

SataQiu replied:
I tested it locally, and it seems that the generated code(`zz_generated.defaults.go`) won't be checked...

pacoxu replied:
It makes sense. So generated code are skipped for that import restrictions.

neolit123 replied:
> 2\. maybe there is already `+foo` directive for it, like `+optional`.

```
~/go/src/k8s.io/kubernetes$ git diff
diff --git a/cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go b/cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go
index 314be0b30c7..38819c814f4 100644
--- a/cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go
+++ b/cmd/kubeadm/app/apis/kubeadm/v1beta4/defaults.go
@@ -74,6 +74,7 @@ func SetDefaults_InitConfiguration(obj *InitConfiguration) {
 }
 
 // SetDefaults_ClusterConfiguration assigns default values for the ClusterConfiguration
+// +k8s:defaulter-gen=covers
 func SetDefaults_ClusterConfiguration(obj *ClusterConfiguration) {
        if obj.KubernetesVersion == "" {
                obj.KubernetesVersion = DefaultKubernetesVersion
```

results in no import of `k8s.io/kubernetes/pkg/api`

> // An existing defaulter method (`SetDefaults_TYPE`) can provide the
> // comment tag:
> //
> //	// +k8s:defaulter-gen=covers
> //
> // to indicate that the defaulter does not or should not call any nested
> // defaulters.

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/code-generator/cmd/defaulter-gen/main.go#L35-L41

neolit123 replied:
result diff:

```
diff --git a/cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go b/cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go
index b36b0f0f952..d2409b0de6a 100644
--- a/cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go
+++ b/cmd/kubeadm/app/apis/kubeadm/v1beta4/zz_generated.defaults.go
@@ -23,7 +23,6 @@ package v1beta4
 
 import (
        runtime "k8s.io/apimachinery/pkg/runtime"
-       v1 "k8s.io/kubernetes/pkg/apis/core/v1"
 )
 
 // RegisterDefaults adds defaulters functions to the given scheme.
@@ -39,41 +38,6 @@ func RegisterDefaults(scheme *runtime.Scheme) error {
 
 func SetObjectDefaults_ClusterConfiguration(in *ClusterConfiguration) {
        SetDefaults_ClusterConfiguration(in)
-       if in.Etcd.Local != nil {
-               for i := range in.Etcd.Local.ExtraEnvs {
-                       a := &in.Etcd.Local.ExtraEnvs[i]
-                       if a.ValueFrom != nil {
-                               if a.ValueFrom.FieldRef != nil {
-                                       v1.SetDefaults_ObjectFieldSelector(a.ValueFrom.FieldRef)
-                               }
-                       }
-               }
-       }
-       SetDefaults_APIServer(&in.APIServer)
-       for i := range in.APIServer.ControlPlaneComponent.ExtraEnvs {
-               a := &in.APIServer.ControlPlaneComponent.ExtraEnvs[i]
-               if a.ValueFrom != nil {
-                       if a.ValueFrom.FieldRef != nil {
-                               v1.SetDefaults_ObjectFieldSelector(a.ValueFrom.FieldRef)
-                       }
-               }
-       }
-       for i := range in.ControllerManager.ExtraEnvs {
-               a := &in.ControllerManager.ExtraEnvs[i]
-               if a.ValueFrom != nil {
-                       if a.ValueFrom.FieldRef != nil {
-                               v1.SetDefaults_ObjectFieldSelector(a.ValueFrom.FieldRef)
-                       }
-               }
-       }

```


questions:
- would that be a problem for us in the future? probably no, because if we want a defaulter for some field in ClusterConfiguration we can explicitly call it in SetDefaults_ClusterConfiguration()
- can this missing defaulting for `a.ValueFrom.FieldRef` be a problem? i think it just sets that field selector to "v1", but technically the kube-apiserver must also do this defaulting when the mirror Pod (of the static pod) is created, right?

chendave replied:
okay, I like this approach, it's more neat to look!
